File: Sliding_Window/Python/121_best_time_to_buy_and_sell_stock.py
# RUNTIME: 25 ms, faster than 89.55% of python3 Submissions. 
# MEMORY USAGE: 26.95 MB, less than 32.87% of python3 Submissions.

from __future__ import annotations
from ast import *

# My thought process:
# 1. assign right pointer to 0 and left pointer to first element
# 2. assign maxVal to 0
# 3. iterate through prices
# 4. right pointer gets current price
# 5. if value at right pointer is higher than left pointer check if maxVal needs to be updated
# 6. otherweise if right pointer is lower than left pointer update left pointer to right value
# 7. continue until all prices are processed
# 8. Return maxVal (if no profit possible this will be 0 as initialized)


# Time complexity is O(n) as we only iterate once through the prices list
# Space complexity is O(1) as no additional data structures dependent on input size are used
class Solution:
    def maxProfit(self, prices: List[int]) -> int:
        right = 0
        left = prices[0]
        maxVal = 0
        
        for current_price in prices:
            right = current_price
            if (right > left):
                if (maxVal < right - left):
                    maxVal = right - left
            elif (right < left):
                left = right

        return maxVal
                

File: Sliding_Window/C++/121_best_time_to_buy_and_sell_stock.cpp
/*
* RUNTIME: 0 ms - Beats 100.00% of submissions
* MEMORY - 97.46 MB - Beats 25.75% of submissions
*/

#include <vector>
#include <string>
#include <algorithm>
#include <set>
#include <stack>
using namespace std;


/* Thoughtprocess:
* 1. Initialize two pointers, left and right.
* 2. left pointer tracks the minimum price so far.
* 3. right pointer iterates through each price in the list.
* 4. If the current price (right) is greater than left, calculate potential profit.
* 5. Update maxValue if the calculated profit is higher than the current maxValue.
* 6. If the current price (right) is less than left, update left to the current price.
* 7. Continue until all prices are processed.
* 8. Return maxValue as the maximum profit achievable.
* 
*/

// Time Complexity: O(n) - single pass through prices
// Space Complexity: O(1) - only a few extra variables used

class Solution {
public:
    int maxProfit(vector<int>& prices) {
        int left = prices[0];
        int right = 0;
        int maxValue = 0;

        for (auto current_price : prices)
        {
            right = current_price;
            if (right > left)
            {
                if (maxValue < right - left)
                {
                    maxValue = right - left;
                }
            }
            else if (right < left)
            {
                left = right;
            }
        }
        return maxValue;
    }
};